### ISSUE TITLE
deadlock when training torch model

### ISSUE THREAD

**Issue by egillax** [mlverse/torch#1044]
Hi @dfalbel ,

I'm having an issue that the model training just hangs when training models. This doesn't happen always but has now happened twice in the last week. I'm using torch 0.10 on ubuntu 20. 

I'm attaching backtraces generated with gdb and `thread apply all bt full` when attaching to the hanging process.

Any idea what could be going on or how to debug further? 

[gdb_deadlock1.txt.txt](https://github.com/mlverse/torch/files/11477059/gdb_deadlock1.txt.txt)
[gdb_deadlock2.txt.txt](https://github.com/mlverse/torch/files/11477060/gdb_deadlock2.txt.txt)

**Comment by dfalbel** [comment]
HI @egillax ,

Is the process running torch forked at some point?
Usually, forking is not safe when using LibTorch, if doing parallel work it's better to use multi-process parallelization. 
If you really need forking, you must make sure that you don't use autograd in the main process (the one that will be forked) before forking, otherwise bad things can happen, incluing deadlocks like this one.

**Comment by egillax** [comment]
Hi @dfalbel,

No there shouldn't be any forking. There should be only one process running the model training. I did notice though that for the first deadlock that happened there was another process from an older rsession occupying gpu memory. But for the later deadlock there was only one process using the gpu, not sure that's relevant.

**Comment by dfalbel** [comment]
This is weird! It's interesting to see some JVM symbols in the backtrace. Do you know where they could come from?

Eg:

```
#7  0x00007fece5a97406 in ?? () from /usr/lib/jvm/java-11-openjdk-amd64/lib/server/libjvm.so
```

It seems that the deadlock situation is very similar to what we saw in #971: Autograd is running and, it tries to allocate more memory, which is not possible because GPU is at it's max, it then tries to cleanup some memory by calling the GC, which in turn tries to release some memory but gets deadlocked.

**Comment by egillax** [comment]
I'm not sure where the jvm stuff is coming from. There are packages earlier in my pipeline using java, to connect to a database and fetch the data.

I'm trying now to reproduce the issue in a simpler setting.  Originally this happened on a server running my full pipeline. Now I'm trying to run only the affected code on the server with same data and separately on my laptop with fake data.

**Comment by egillax** [comment]
Yes the code is public. The main training loop is in [this class](https://github.com/OHDSI/DeepPatientLevelPrediction/blob/main/R/Estimator-class.R#L104) which is instantiated and fit during hyperparameter tuning [here](https://github.com/OHDSI/DeepPatientLevelPrediction/blob/main/R/Estimator.R#L273). The issue has both happened with a [ResNet](https://github.com/OHDSI/DeepPatientLevelPrediction/blob/main/R/ResNet.R) and a [Transformer](https://github.com/OHDSI/DeepPatientLevelPrediction/blob/main/R/Transformer.R).

I'm also trying to make a more minimal example I can share, will post it here if I manage.
